logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/backup/RestoreBackupUseCase.kt

```diff
@@ -1,38 +1,121 @@
+import com.wire.kalium.logic.kaliumLogger
+import com.wire.kalium.logic.util.extractCompressedFile
 import com.wire.kalium.logic.wrapStorageRequest
-import com.wire.kalium.persistence.backup.BackupImporter
+import com.wire.kalium.persistence.backup.DatabaseImporter
 import com.wire.kalium.util.KaliumDispatcher
 import com.wire.kalium.util.KaliumDispatcherImpl
 import kotlinx.coroutines.withContext
+import kotlinx.serialization.decodeFromString
+import kotlinx.serialization.json.Json
+import okio.Path
+import okio.Source
+import okio.buffer
+import okio.use
 
 interface RestoreBackupUseCase {
 
     /**
      * Restores a valid previously created backup file into the current database, respecting the current data if there is any overlap.
-     * @param backUpFilePath The absolute file system path to the backup file
+     * @param extractedBackupRootPath The absolute file system path to the backup file.
+     * @param password the password used to encrypt the original backup file. Null if the file was not encrypted.
+     * @return A [RestoreBackupResult] indicating the success or failure of the operation.
      */
-    suspend operator fun invoke(backupFilePath: String): RestoreBackupResult
+    suspend operator fun invoke(extractedBackupRootPath: Path, password: String?): RestoreBackupResult
 }
 
 internal class RestoreBackupUseCaseImpl(
-    private val backupImporter: BackupImporter,
-    private val dispatchers: KaliumDispatcher = KaliumDispatcherImpl
+    private val databaseImporter: DatabaseImporter,
+    private val kaliumFileSystem: KaliumFileSystem,
+    private val userId: UserId,
+    private val dispatchers: KaliumDispatcher = KaliumDispatcherImpl,
 ) : RestoreBackupUseCase {
 
-    override suspend operator fun invoke(backupFilePath: String): RestoreBackupResult {
-        return withContext(dispatchers.io) {
-            wrapStorageRequest {
-                backupImporter.importFromFile(backupFilePath)
-            }.fold({ RestoreBackupResult.Failure(it) }, { RestoreBackupResult.Success })
+    override suspend operator fun invoke(extractedBackupRootPath: Path, password: String?): RestoreBackupResult =
+        withContext(dispatchers.io) {
+            val idMapper = MapperProvider.idMapper()
+            val encryptedFilePath = kaliumFileSystem.list(extractedBackupRootPath).firstOrNull { it.name.contains(".cc20") }
+            val isPasswordProtected = !password.isNullOrEmpty()
+
+            // Check if the backup file is encrypted and if the password is provided
+            if (encryptedFilePath != null && !isPasswordProtected) return@withContext RestoreBackupResult.Failure(
+                DecryptionFailure("A non blank password is required to restore a password protected backup")
+            )
+            if (encryptedFilePath == null && isPasswordProtected) return@withContext RestoreBackupResult.Failure(
+                DecryptionFailure("A password was provided but no encrypted backup file was found")
```

Comment:
do we care about non null encrypedFilepath if it is not passowrd protected ? the same goes for carrying about isPasswordProtected when encryptedFilePath == null